src/main/java/io/jsonwebtoken/impl/DefaultJwtParser.java

```diff
@@ -342,6 +350,14 @@ public Jwt parse(String jwt) throws ExpiredJwtException, MalformedJwtException,
             //re-create the jwt part without the signature.  This is what needs to be signed for verification:
             String jwtWithoutSignature = base64UrlEncodedHeader + SEPARATOR_CHAR + base64UrlEncodedPayload;
 
+            if (signatureAlgorithm != null &&
+                    (signatureAlgorithm.compareTo(algorithm) != 0)) {
+                String message = "The expected algorithm " + signatureAlgorithm.getValue() + " does not match " +
+                        "the algorithm specified in the JWT header (" + algorithm.getValue() + "). There is a " +
+                        "chance this JWT is trying to force algorithm selection.";
+                throw new SignatureException(message);
```

Comment:
I'm not sure `SignatureException` is valid here - ordinarily that was intended to be used during actual signature calculation and verification operations, not header comparisons.  I think `IncorrectHeaderException` (paralleling the `IncorrectClaimException`) might be better.

In other words, just as we have:
```
ClaimJwtException
|-- InvalidClaimException
    |-- MissingClaimException
    |-- IncorrectClaimException
```
we might also need:
```
HeaderJwtException
|-- InvalidHeaderException
    |-- MissingHeaderException
    |-- IncorrectHeaderException
```